theano/sandbox/cuda/acdc.py

```diff
@@ -0,0 +1,309 @@
+#
+# The following is based on acdc_wrapper.cu
+# from https://github.com/mdenil/acdc-torch
+#
+
+import os
+
+import theano
+from theano.sandbox.cuda import GpuOp
+from theano.sandbox.cuda.basic_ops import as_cuda_ndarray_variable, gpu_contiguous
+
+import cuda_ndarray
+
+class FastACDC(GpuOp):
+
+    def c_headers(self):
+        return ['<stdio.h>', '<cufft.h>']
+
+    def c_libraries(self):
+        return ['cufft']
+
+    def make_node(self, x, A, Ab, D, Db):
+        x_ = gpu_contiguous(as_cuda_ndarray_variable(x))
+        A_ = gpu_contiguous(as_cuda_ndarray_variable(A))
+        Ab_ = gpu_contiguous(as_cuda_ndarray_variable(Ab))
+        D_ = gpu_contiguous(as_cuda_ndarray_variable(D))
+        Db_ = gpu_contiguous(as_cuda_ndarray_variable(Db))
+        for t in [x_, A_, Ab_, D_, Db_]:
+            assert t.dtype == 'float32'
+        return theano.Apply(self, [x_, A_, Ab_, D_, Db_], [x_.type()])
+
+    def grad(self, inputs, cost_grad):
+        inp, A, Ab, D, Db = inputs
+        gradOutput, = cost_grad
+        gradOutput = gpu_contiguous(gradOutput)
+        return FastACDCGrad()(inp, gradOutput, A, Ab, D, Db)
+
+    def c_support_code_apply(self, node, nodename):
+        # Based on a similar mechanism in sandbox/cuda/blas.py
+        #
+        # REMEMBER TO RAISE c_code_cache_version when changing any of
+        # these files
+        files = ['acdc.cu']
+        codes = [open(os.path.join(os.path.split(__file__)[0], f)).read()
+                for f in files]
+        return reduce(str.__add__, codes)
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        x, A, Ab, D, Db = inputs
+        z, = outputs
+        fail = sub['fail']
+        return """
+{
+    // Validate that the output storage exists.
+    if (NULL == %(z)s ||
+        CudaNdarray_NDIM(%(z)s) != CudaNdarray_NDIM(%(x)s))
```

Comment:
The output can be of different shape as the user can call Theano function with inputs of different shape to different call. If the output have the same shape as the input as in this case, you can use this method CudaNdarray_prep_output() that will do all that for you.